### PR TITLE
Clean built artifacts after building extensions

### LINKED ISSUE #3958: Cleanup object files on gems with extensions
I've just noticed that gems with compiled extensions use a lot of disk space, most of it are object files that, as far as I understand, are not needed on runtime.

For example, on sassc-2.4.0 does use (on my linux computer) 114.3 MiB, where extension code is 110.9 MiB but the important files (as far as I understand) are just `libsass`, with 1,7 MB and `libsass.so` 3,2MB. That is about 108 MB of extra space that could also make slower deployments, when generating bundles and testing when caching gems on CI, or simply on developers machines that duplicate gems on different environments (per project GEM_HOME).

Here are my current environment details:

```
$ gem env version 3.0.3
```

What are your thoughts on this?

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
## What was the end-user or developer problem that led to this PR?

Currently installing extensions leaves around intermediary files that may end up taking a lot of disk space.

## What is your fix for the problem, implemented in this PR?

Add a clean step after installing the extension.

This PR is based on the useful work by @eloyesp at #4017.

Fixes #3958.

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/bundler/doc/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

File: lib/rubygems/ext/builder.rb [hunk 1]
```diff
@@ -17,7 +17,7 @@ def self.class_name
     $1.downcase
   end
 
-  def self.make(dest_path, results, make_dir = Dir.pwd, sitedir = nil)
+  def self.make(dest_path, results, make_dir = Dir.pwd, targets = ["clean", "", "install"])
```
Comment (eloyesp) [inline lib/rubygems/ext/builder.rb:20]: I don't know what `sitedir` mean... is it being used?

Anyway, I would not modify the signature of the method this way, may be adding a keyword argument to make sure to prevent misunderstandings.

File: lib/rubygems/ext/ext_conf_builder.rb [hunk 2]
```diff
@@ -39,35 +31,22 @@ def self.build(extension, dest_path, results, args=[], lib_dir=nil, extension_di
 
       ENV["DESTDIR"] = nil
 
-      make dest_path, results, extension_dir, tmp_dest_relative
-
-      full_tmp_dest = File.join(extension_dir, tmp_dest_relative)
+      rel_dest_path = Pathname.new(dest_path).relative_path_from(Pathname.new(extension_dir))
```
Comment (eloyesp) [inline lib/rubygems/ext/ext_conf_builder.rb:34]: I prefer the `Pathname(dest_path)` syntax, without `.new`, it's cleaner and easier to read.

**Review by eloyesp — COMMENTED** [review]
I like these simplifications a lot, just wanted to add that there might be a secondary bug that was hidden by the build artifacts. 

That is: when bundle install a git sourced gem, it adds some non-existing paths to the load path, I've added a check for that on `Bundler::Runtime#setup` breaking multiple escenarios.

The problem is that I'm not entirely sure how rubygems and bundler interact when building extensions on git sourced gems.

Hope my comments help somehow.

File: lib/rubygems/ext/builder.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply lib/rubygems/ext/builder.rb]: The `sitedir` parameter was only introduced very recently at https://github.com/rubygems/rubygems/commit/dea41fa2dcce8a6c11c7c92c73db7a66a7d7ed51. I'm aware our extension building code is quite complicated, but it used to be worse 😅.

The above change was released with RubyGems 3.3.17, quite recently, so I don't think removing changing it with another positional parameter is going to break anything.

That said, I want to get to a green base line first, then figure out the details.

File: lib/rubygems/ext/ext_conf_builder.rb (hunk 2 above)
Comment (deivid-rodriguez) [reply lib/rubygems/ext/ext_conf_builder.rb]: I actually find the more explicit `Pathname.new` easier to pase, I guess because it's more common Ruby. And it's also the style we generally use in the rest of the codebase. That said, I was also considering to not introduce a dependency on `pathname` just for this, and instead modify the existing `get_relative_path` helper. Will figure this out later.

**Comment by deivid-rodriguez** [comment]
> That is: when bundle install a git sourced gem, it adds some non-existing paths to the load path, I've added a check for that on `Bundler::Runtime#setup` breaking multiple escenarios.

Yeah, building extensions for git gems is a big hack at the moment, and it indeed has the consequence of some paths to be added to the `$LOAD_PATH` that don't exist. That's actually what prevents us to refactor things further and also stop copying build artifacts to `lib_dir`, because that's apparently what makes extensions for `git` gems actually work.

We should refactoring things further to just be able to `make, make install, make clean` to install extensions, no matter the source. But it sounds too much for this PR.

File: lib/rubygems/ext/builder.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply lib/rubygems/ext/builder.rb]: This refactorings broke Windows 2.6 and Windows 2.7 for some reason. I really don't want to investigate that so I removed those simplifications and will be opening a separate PR that can be merged by the end of next year when we will drop Ruby 2.6 and Ruby 2.7 support.

**Comment by deivid-rodriguez** [comment]
After investigating the issues here, I think this is safe to do.

I could only see this breaking if someone is intentionally changing the `$LOAD_PATH` to point to where the artifacts live (normally the `ext/` folder of an installed gem). These artifacts are already installed in two places right now (the `lib` folder of the installed gem, and the `extensions` folder), and both are already in the `$LOAD_PATH` by default. So it makes no sense that someone did this, and if they did, it can't hardly be considered a breaking change in my opinion. Also, cleanup is currently limited to `extconf` extension builder, so the scope is even more limited.

So from my side this is ready.

**Comment by eloyesp** [comment]
Just one more question, will this impact bundler installs? Is bundler installing  gems using the `extconf` builder or is it running his own builder?

**Comment by deivid-rodriguez** [comment]
Yes, Bundler delegates to RubyGems to install extensions, so it will use whatever builder each gem uses. If gems define their extension through an `extconf.rb` file, then Bundler will use that builder from RubyGems.

**Comment by simi** [comment]
Thanks for digging in this @deivid-rodriguez. I did other review of the build artifact handling. I agree this seems safe. I did quick test to see impact on sassc issue https://github.com/sass/sassc-ruby/issues/200 and bundle size went from 65MB to 25MB.

**Comment by simi** [comment]
Thanks @eloyesp for pushing this forward!

**Comment by lipidity** [comment]
Hi 👋 

Running `make clean` after installing an extension breaks https://github.com/ohler55/ox because it removes the built artifacts which are needed at runtime.

**Comment by simi** [comment]
> 

https://github.com/rubygems/rubygems/issues/6205#issuecomment-1365622769

**Comment by simi** [comment]
> Hi wave
> 
> Running `make clean` after installing an extension breaks https://github.com/ohler55/ox because it removes the built artifacts which are needed at runtime.

https://github.com/rubygems/rubygems/issues/6205#issuecomment-1365622769

**Comment by hsbt** [comment]
`appsignal-ruby` was fixed at https://github.com/appsignal/appsignal-ruby/pull/907.

**Comment by lipidity** [comment]
@simi thanks, I should have linked to https://github.com/ohler55/ox/issues/303

**Comment by hsbt** [comment]
@lipidity No. It works. Please confirm [my comment](https://github.com/ohler55/ox/issues/303#issuecomment-1370506686)

**Comment by lipidity** [comment]
I was wrong about ox still being broken. Confirmed that it works.
